### PR TITLE
SonarCloud in Security Workflows

### PR DESCRIPTION
**Pull request by christophe-havard-sonarsource**
## Pre-requisites

- [ ] Prior to submitting a new workflow, please apply to join the GitHub Technology Partner Program: [partner.github.com/apply](https://partner.github.com/apply?partnershipType=Technology+Partner).

---

### **Please note that at this time we are only accepting new starter workflows for Code Scanning. Updates to existing starter workflows are fine.**

---

## Tasks

**For _all_ workflows, the workflow:**

- [ ] Should be contained in a `.yml` file with the language or platform as its filename, in lower, [_kebab-cased_](https://en.wikipedia.org/wiki/Kebab_case) format (for example, [`docker-image.yml`](https://github.com/actions/starter-workflows/blob/main/ci/docker-image.yml)).  Special characters should be removed or replaced with words as appropriate (for example, "dotnet" instead of ".NET").
- [ ] Should use sentence case for the names of workflows and steps (for example, "Run tests").
- [ ] Should be named _only_ by the name of the language or platform (for example, "Go", not "Go CI" or "Go Build").
- [ ] Should include comments in the workflow for any parts that are not obvious or could use clarification.
- [ ] Should specify least priviledge [permissions](https://docs.github.com/en/actions/security-guides/automatic-token-authentication#modifying-the-permissions-for-the-github_token) for `GITHUB_TOKEN` so that the workflow runs successfully. 

**For _CI_ workflows, the workflow:**

- [ ] Should be preserved under [the `ci` directory](https://github.com/actions/starter-workflows/tree/main/ci).
- [ ] Should include a matching `ci/properties/*.properties.json` file (for example, [`ci/properties/docker-publish.properties.json`](https://github.com/actions/starter-workflows/blob/main/ci/properties/docker-publish.properties.json)).
- [ ] Should run on `push` to `branches: [ $default-branch ]` and `pull_request` to `branches: [ $default-branch ]`.
- [ ] Packaging workflows should run on `release` with `types: [ created ]`.
- [ ] Publishing workflows should have a filename that is the name of the language or platform, in lower case, followed by "-publish" (for example, [`docker-publish.yml`](https://github.com/actions/starter-workflows/blob/main/ci/docker-publish.yml)).

**For _Code Scanning_ workflows, the workflow:**

- [ ] Should be preserved under [the `code-scanning` directory](https://github.com/actions/starter-workflows/tree/main/ci).
- [ ] Should include a matching `code-scanning/properties/*.properties.json` file (for example, [`code-scanning/properties/codeql.properties.json`](https://github.com/actions/starter-workflows/blob/main/code-scanning/properties/codeql.properties.json)), with properties set as follows:
  - [ ] `name`: Name of the Code Scanning integration.
  - [ ] `organization`: Name of the organization producing the Code Scanning integration.
  - [ ] `description`: Short description of the Code Scanning integration.
  - [ ] `categories`: Array of languages supported by the Code Scanning integration.
  - [ ] `iconName`: Name of the SVG logo representing the Code Scanning integration. This SVG logo must be present in [the `icons` directory](https://github.com/actions/starter-workflows/tree/main/icons).
- [ ] Should run on `push` to `branches: [ $default-branch, $protected-branches ]` and `pull_request` to `branches: [ $default-branch ]`. We also recommend a `schedule` trigger of `cron: $cron-weekly` (for example, [`codeql.yml`](https://github.com/actions/starter-workflows/blob/c59b62dee0eae1f9f368b7011cf05c2fc42cf084/code-scanning/codeql.yml#L14-L21)).

**Some general notes:**

- [ ] This workflow must _only_ use actions that are produced by GitHub, [in the `actions` organization](https://github.com/actions), **or**
- [ ] This workflow must _only_ use actions that are produced by the language or ecosystem that the workflow supports.  These actions must be [published to the GitHub Marketplace](https://github.com/marketplace?type=actions).  We require that these actions be referenced using the full 40 character hash of the action's commit instead of a tag.  Additionally, workflows must include the following comment at the top of the workflow file:
    ```
    # This workflow uses actions that are not certified by GitHub.
    # They are provided by a third-party and are governed by
    # separate terms of service, privacy policy, and support
    # documentation.
    ```
- [ ] Automation and CI workflows should not send data to any 3rd party service except for the purposes of installing dependencies.
- [ ] Automation and CI workflows cannot be dependent on a paid service or product.

### CODE REVIEW

**Comment by bishal-pdMSFT** [comment]
@josepalafox can you please review this new starter workflow?

File: code-scanning/properties/sonarcloud.properties.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+    "name": "SonarCloud",
+    "creator": "Sonar",
+    "description": "Static analysis of code for vulnerability detection, covering 26+ languages. Free for open source projects",
```
Comment (bishal-pdMSFT) [inline code-scanning/properties/sonarcloud.properties.json:4]:
```suggestion
    "description": "Static analysis of code for vulnerability detection, covering 26+ languages.",
```

I would refrain from using `free`/`paid` or talking about plans here. You can do that by putting a comment in workflow file.

Comment (bishal-pdMSFT) [reply code-scanning/properties/sonarcloud.properties.json:4]: Rather call out any quick USP which Sonar  cloud may provide.

File: code-scanning/sonarcloud.yml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+jobs:
+  Analysis:
+    runs-on: ubuntu-latest
+    
+    steps:
+      - name: Analyze with SonarCloud         
+
+          # You can pin the exact commit or the version.
+          # uses: SonarSource/sonarcloud-github-action@de2e56b42aa84d0b1c5b622644ac17e505c9a049
+          uses: SonarSource/sonarcloud-github-action@v1.6
```
Comment (bishal-pdMSFT) [inline code-scanning/sonarcloud.yml:46]: By default pin to commit SHA

File: code-scanning/sonarcloud.yml [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+          env:
+            GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}  # Needed to get PR information, if any
```
Comment (bishal-pdMSFT) [inline code-scanning/sonarcloud.yml:48]: If you are using `GITHUB_TOKEN`, please ensure you have provided least privileged [permission](https://docs.github.com/en/actions/security-guides/automatic-token-authentication#modifying-the-permissions-for-the-github_token) explicitly.

File: code-scanning/sonarcloud.yml (hunk 2 above)
Comment (christophe-havard-sonarsource) [reply code-scanning/sonarcloud.yml:46]: you mean, the SHA of the very last commit?

File: code-scanning/sonarcloud.yml (hunk 3 above)
Comment (christophe-havard-sonarsource) [reply code-scanning/sonarcloud.yml:49]: Do you mean that we need to ensure that in our documentation or put a word directly in this file?

File: code-scanning/sonarcloud.yml (hunk 2 above)
Comment (Phantsure) [reply code-scanning/sonarcloud.yml:50]: Yes, the commit sha for this release

File: code-scanning/properties/sonarcloud.properties.json (hunk 1 above)
Comment (Phantsure) [reply code-scanning/properties/sonarcloud.properties.json:4]: @christophe-havard-sonarsource Can you please address this as well? Thanks

File: code-scanning/sonarcloud.yml (hunk 2 above)
Comment (Phantsure) [reply code-scanning/sonarcloud.yml:50]: I believe this is the sha: de2e56b42aa84d0b1c5b622644ac17e505c9a049

File: code-scanning/sonarcloud.yml [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+jobs:
+  Analysis:
+    runs-on: ubuntu-latest
+    
+    steps:
+      - name: Analyze with SonarCloud         
+
+          # You can pin the exact commit or the version.
+          # uses: SonarSource/sonarcloud-github-action@de2e56b42aa84d0b1c5b622644ac17e505c9a049
+          uses: SonarSource/sonarcloud-github-action@v1.6
```
Comment (Phantsure) [inline code-scanning/sonarcloud.yml:50]:
```suggestion
          uses: SonarSource/sonarcloud-github-action@de2e56b42aa84d0b1c5b622644ac17e505c9a049
```

**Review by Phantsure — CHANGES_REQUESTED** [review]

File: code-scanning/sonarcloud.yml (hunk 4 above)
Comment (Phantsure) [reply code-scanning/sonarcloud.yml:50]: I believe this is correct sha

File: code-scanning/properties/sonarcloud.properties.json (hunk 1 above)
Comment (christophe-havard-sonarsource) [reply code-scanning/properties/sonarcloud.properties.json]: done

**Comment by Phantsure** [comment]
@christophe-havard-sonarsource can you look into this check failing?

**Comment by christophe-havard-sonarsource** [comment]
@Phantsure everything has been fixed, it's good to go ;)

**Review by bishal-pdMSFT — APPROVED** [review]

**Comment by bishal-pdMSFT** [comment]
@Phantsure this looks good to merge!

**Review by Phantsure — APPROVED** [review]

**Pull request merged**